lib/vfscore/main.c

```diff
@@ -1167,36 +1182,85 @@ int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
 	int error;
 	struct vfscore_file *fp;
 
+	/**
+	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
+	 * a dirent64 structure instead of the `struct dirent` entry parameter.
+	 */
+	struct dirent64 entry64;
+
 	trace_vfs_readdir(dir->fd, entry);
 	error = fget(dir->fd, &fp);
 	if (!error) {
-		error = sys_readdir(fp, entry);
+		error = sys_readdir(fp, &entry64);
 		fdrop(fp);
-		if (error) {
+		if (unlikely(error))
 			trace_vfs_readdir_err(error);
-		} else {
+		else
 			trace_vfs_readdir_ret();
-		}
 	}
 	// Our dirent has (like Linux) a d_reclen field, but a constant size.
 	entry->d_reclen = sizeof(*entry);
+	entry->d_ino = entry64.d_ino;
```

Comment:
Is this really fine :thinking:? You are referencing the fields of `entry64`, which may have not been initialized unless we entered the `if` statement block above, shouldn't you probably do this in the if block statement above :thinking:? Also this whole function looks like it's asking to be re-ordered. that `if` statement block looks... unnecessarily imbricated/large.